bayes_kit/drghmc.py

```diff
@@ -0,0 +1,437 @@
+
+        Acceptance probability for the k-th proposed sample, with probabilistic delayed
+        rejection, is computed by extending eqns. 29, 30, and 31 of Modi et al. The
+        acceptance probability contains three terms for the current and proposed
+        samples respectively:
+
+            1. the joint log density
+            2. the hastings term for asymmetric proposal kernels; simplfies to the probability of rejecting previous/ghost proposals
+            3. the probability of proposing another sample upon rejection
+
+        Optimize computation by passing hastings_cur and logp from previous proposal
+        and by returning logp_prop. This is less readable but ensures (1) only one
+        density evaluation per sample outside of leapfrog, (2) no recursion is required
+        to compute the denominator of eqn. 29, and (3) recursion on the numerator of
+        eqn. 29 is done efficiently.
+
+        Proposed samples are generated by running the leapfrog integrator from the
+        *current sample*. To maintain detailed balance, 'ghost proposals' are similarly
+        generated by running the leapfrog integrator from the *proposed sample*.
+        (Proposed samples and ghost samples of the same order are generated with the
+        same leapfrog stepsize and number of steps.)
+
+        When acceptance probability of a ghost proposal is 1, the hastings term of the
+        proposed sample is completely zeroed out. Thus early stopping is implemented to
+        avoid extra computation and sidestep negative infinity log probabilities.
+
+        Flip momentum after each ghost proposal to ensure detailed balance.
+
+        Args:
+            theta_prop: proposed position
+            rho_prop: proposed momentum
+            k: number of delayed rejection proposals
+            hastings_cur: log probability of rejecting all previous proposals
+            logp: log probability of current sample
+
+        Returns:
+            tuple[float, float]: log probability of acceptance and proposed sample
+        """
+        logp_prop = self.joint_logp(theta_prop, rho_prop)
+        hastings_prop = 0
+
+        for i in range(k):
+            stepsize, steps = self._stepsize_list[i], self._steps_list[i]
+            theta_ghost, rho_ghost = self.leapfrog(
+                theta_prop, rho_prop, stepsize, steps
+            )
+            rho_ghost = -rho_ghost
+
+            accept_logp, _ = self.accept(
+                theta_ghost, rho_ghost, i, hastings_prop, logp_prop
+            )
+            if accept_logp == 0:  # early stopping to avoid -inf in np.log1p
+                return -np.inf, logp_prop
+            reject_logp = np.log1p(-np.exp(accept_logp))
+            hastings_prop += reject_logp
+
+        pdr_prop = self._prob_of_delayed_rejection(hastings_prop)
+        pdr_cur = self._prob_of_delayed_rejection(hastings_cur)
+
+        detailed_balance = (
```

Comment:
Detailed balance is a property of proposal distributions at equilibrium.  It's not a number, so you shouldn't be using it as a variable name here.  Is this the acceptance log probability, perhaps?